File: python-refactor/Code/pre_03_station/CN_station_03_read_raw_outlier.m
clear all; close all; clc

tic
% 
% path_data = '//10.72.26.56/irisnas5/Data/';

path_data = '/share/irisnas5/Data/';
path = '/share/irisnas5/Data/Station/Station_CN/';
% addpath(genpath('/share/irisnas5/Data/matlab_func/'))

stn_info_cn = csvread([path_data,'Station/Station_CN/cn_stn_code_lonlat_period.csv'],1);
% scode1, scode2, lon, lat, op_start, op_end

%% read files 서희가 만든 china stn 파일 불러와서 년도별로 파일 묶기
% header_ndata = {'doy','yr','mon','day','CST','AQI','PM25','PM25_24h',...
%     'PM10','PM10_24h','SO2','SO2_24h','NO2','NO2_24h','O3','O3_24h',...
%     'O3_8h','O3_8h_24h','CO','CO_24h','scode'};
% 
% for yr=2015:2016
%     cd([path_data,'Station_CN/rm_outlier/',num2str(yr)])
%     list = dir('*.mat');
%     list = {list.name}';
%     
%     ndata = [];
%     for k = 1:length(list)
%         load(list{k}) %stn_CN
%         ndata = [ndata; stn_CN];
%         disp(list{k})
%     end
%     save([path_data,'Station_CN/stn_code_data/stn_code_data_rm_outlier_',num2str(yr)],'ndata','header_ndata','-v7.3')
% end

% list15=dir('stn_code_data_rm_*2015*');
% list15={list15.name}';
% ndata = importdata(list15{1});
% for k=2:length(list15)
%     load(list15{k})
%     ndata=[ndata;stn_CN];
% end
% save([path_data,'Station_CN/stn_code_data/stn_code_data_rm_outlier_',num2str(yr)],'ndata','header_ndata','-v7.3')
% 
% list16=dir('stn_code_data_rm_*2016*');
% list16={list16.name}';
% ndata = importdata(list16{1});
% for k=2:length(list16)
%     load(list16{k})
%     ndata=[ndata;stn_CN];
% end
% save([path_data,'Station_CN/stn_code_data/stn_code_data_rm_outlier_',num2str(yr)],'ndata','header_ndata','-v7.3')

%% stn_scode_data for China
header_ndata = {'doy','yr','mon','day','CST','AQI','PM25','PM25_24h',...
    'PM10','PM10_24h','SO2','SO2_24h','NO2','NO2_24h','O3','O3_24h',...
    'O3_8h','O3_8h_24h','CO','CO_24h','scode','scode2'};

for yr=2019 %2015:2019
ndata=importdata([path_data,'Station/Station_CN/stn_code_data/stn_code_data_rm_outlier_',num2str(yr),'.mat']);
ndata(:,end+1)=0;

ndata_scode = [];
% Assign scode2
for j=1:size(stn_info_cn,1)
    ndata_temp = ndata(ndata(:,end-1)==stn_info_cn(j,1),:);
    for k = 1:5 %%%%%%%%%%%%%%%%%%
        ndata_temp2 = ndata_temp(ndata_temp(:,3)==k,:);
        if isempty(ndata_temp2)==0
            yrmon = yr*100+k;
            idx = (stn_info_cn(j,5)<=yrmon)&(stn_info_cn(j,6)>=yrmon);
            if idx==1
                ndata_temp2(:,end)=stn_info_cn(j,2);
                ndata_scode=[ndata_scode;ndata_temp2];
            end
        end
    end
    if mod(j,100)==0
        save([path,'stn_scode_data_',num2str(yr),'_',num2str(j-99,'%04i'),'.mat'],'ndata_scode','-v7.3')
        ndata_scode = [];
    elseif j==size(stn_info_cn,1)
        save([path,'stn_scode_data_',num2str(yr),'_1501.mat'],'ndata_scode','-v7.3')
    end

    disp([num2str(j),'/',num2str(size(stn_info_cn,1))])
end

ndata_scode = [];
for k = 1:100:1501
    ndata_scode_temp = importdata([path,'stn_scode_data_',num2str(yr),'_',num2str(k,'%04i'),'.mat']);
    ndata_scode = [ndata_scode; ndata_scode_temp];
end
save([path,'stn_scode_data/cn_stn_scode_data_rm_outlier_',num2str(yr)],'ndata_scode','header_ndata','-v7.3')

disp(yr)
end

toc

File: python-refactor/Code/pre_03_station/CN_station_03_read_raw_outlier.py
### Package Import
import sys
import os
base_dir = os.environ['GEMS_HOME']
#base_dir = 'D:\github\GEMS_python'
project_path = os.path.join(base_dir, 'python-refactor')
sys.path.insert(0, project_path)
from Code.utils import matlab

import copy
import numpy as np
import pandas as pd
import glob
import h5py

### Setting path
#data_base_dir = os.path.join('/data2', 'sehyun', 'Data')
#path_station = os.path.join(data_base_dir, 'Preprocessed_raw', 'Station') 

data_base_dir = os.path.join('/', 'share', 'irisnas5', 'GEMS', 'GEMS_python')
#data_base_dir = os.path.join('//', '10.72.26.56','irisnas5', 'GEMS', 'GEMS_python')
path_station = os.path.join(data_base_dir, 'Preprocessed_raw', 'Station') 

# scode1, scode2, lon, lat, op_start, op_
stn_info_cn = pd.read_csv(os.path.join(path_station, 'Station_CN', 'cn_stn_code_lonlat_period.csv'))
stn_info_cn = stn_info_cn.values
## stn_scode_data for China
header_ndata = np.array(['doy','yr','mon','day','CST','AQI','PM25','PM25_24h',
    'PM10','PM10_24h','SO2','SO2_24h','NO2','NO2_24h','O3','O3_24h',
    'O3_8h','O3_8h_24h','CO','CO_24h','scode','scode2'], dtype=h5py.string_dtype(encoding='utf-8'))

YEARS = [2016] # range(2015, 2019+1)
for yr in YEARS:
    fname = f'stn_code_data_rm_outlier_{yr}.mat'
    ndata = matlab.loadmat(os.path.join(path_station,'Station_CN', 'stn_code_data', fname))['stn_CN']
    ndata = np.hstack([ndata, np.zeros([ndata.shape[0], 1])])
    
    ndata_scode = None
    # Assign scode2
    for j in range(stn_info_cn.shape[0]):
        ndata_temp = ndata[ndata[:,-2]==stn_info_cn[j,0],:]

        if yr==2019:
            mm = 5
        else:
            mm = 12
        for k in range(1,mm+1): 
            ndata_temp2 = ndata_temp[ndata_temp[:,2]==k,:]
            if len(ndata_temp2)!=0:
                yrmon = yr*100+k
                idx = (stn_info_cn[j,4]<=yrmon)&(stn_info_cn[j,5]>=yrmon)
                if idx: # True
                    ndata_temp2[:,-1]=stn_info_cn[j,1]
                    if ndata_scode is None:
                        ndata_scode = ndata_temp2
                    else:
                        ndata_scode= np.vstack([ndata_scode,ndata_temp2])
        if ((j+1)%100)==0:
            fname = f'stn_scode_data_{yr}_{j+1-99:04d}.mat'
            matlab.savemat(os.path.join(path_station,'Station_CN', fname), {'ndata_scode':ndata_scode})
            ndata_scode = None
        elif (j+1)==stn_info_cn.shape[0]:
            if ndata_scode is not None:
                fname = f'stn_scode_data_{yr}_{np.int((j+1)/100)*100+1:04d}.mat'
                matlab.savemat(os.path.join(path_station,'Station_CN', fname), {'ndata_scode':ndata_scode})
        print (f'{j+1} / {stn_info_cn.shape[0]}')
    
    ndata_scode = None
    file_list = glob.glob(os.path.join(path_station,'Station_CN', f'stn_scode_data_{yr}_*.mat'))
    file_list.sort()
    for fname in file_list:
        print (f'Reading ... {os.path.basename(fname)}')
        ndata_scode_temp = matlab.loadmat(fname)['ndata_scode']
        print (ndata_scode_temp.shape)
        if ndata_scode is None:
            ndata_scode = ndata_scode_temp
        else:
            ndata_scode = np.vstack([ndata_scode, ndata_scode_temp])
    fname = f'cn_stn_scode_data_rm_outlier_{yr}.mat'
    matlab.savemat(os.path.join(path_station,'Station_CN','stn_scode_data', fname), {'ndata_scode':ndata_scode})
    with h5py.File(os.path.join(path_station,'Station_CN','stn_scode_data', fname), 'a') as dst:
        dst['header_ndata'] = header_ndata
    print (yr)